src/librustc/ty/context.rs

```diff
@@ -315,6 +315,17 @@ impl<'a, V> LocalTableInContextMut<'a, V> {
     }
 }
 
+/// All information necessary to validate and reveal an `impl Trait` or `existential Type`
+#[derive(RustcEncodable, RustcDecodable, Debug)]
+pub struct ResolvedOpaqueTy<'tcx> {
+    /// The revealed type as seen by this function.
+    pub concrete_type: Ty<'tcx>,
+    /// Generic parameters on the opaque type as passed by this function.
+    /// For `existential type Foo<A, B>; fn foo<T, U>() -> Foo<T, U> { .. }` this is `[T, U]`, not
+    /// `[A, B]`
+    pub substs: &'tcx Substs<'tcx>,
```

Comment:
Hang on, isn't substs checking handled by `wfcheck`? https://github.com/rust-lang/rust/blob/5574b1df577f737373b42cbf364b6cab2dfa5960/src/librustc_typeck/check/wfcheck.rs#L818-L837

Wouldn't that make `ResolvedOpaqueTy` and most of this PR entirely redundant, if params were checked there?

Reply:
Hmm, I can't remove `ResolvedOpaqueTy` as it seems MIR borrowck needs the `Substs`, and I'd rather not touch that code if I can help it.